### ISSUE TITLE
Bi-LSTM self-attention setting(LSTM_selfAttention branch)

### ISSUE THREAD

**Issue by skkulsc** [skkulsc/text#4]
Classification_LSTM_selfAttention Model setting --- 2018-12-04 08:04:50.812867

for_testing = [ False ]

trainingDataPath = [ ../data/merged_dataset/train.txt ]

validationDataPath = [ ../data/merged_dataset/valid.txt ]

testingDataPath = [ ../data/merged_dataset/test.txt ]

wordDicPath = [ ../data/word_embedding_100000/wordDictionary_100000.pkl ] - vocab_size = [ 100000]

Model Training Epochs = [   300]

batch_size = [ 512] - update_size = [ 512] - training_step = [   1]

sequence_length = [ 200] - category_size = [ 6] - embedding_size = [ 512], embedding_trainable = [True]

train_X.shape = [ (520601, 200) ] - valid_X.shape = [ (142510, 200) ]

lstm_hidden_size = [  32] - attention_sizes = [ [128, 64] ] - penalty_coef = [ 0.02000] - fc_output = [ 128] 

lr = [ 0.00100] - lr_decay = [ 0.97000] - fc_dropout_keep_prob = [0.50] - emb_dropout_keep_prob = [0.50]

train_X.shape = [ (520601, 200) ] - valid_X.shape = [ (142510, 200) ]

save_folder = [ ./Classification_LSTM_selfAttention/2018-12-04__08:02:16 ]

save_check_point_name = [ ./Classification_LSTM_selfAttention/2018-12-04__08:02:16/model/Classification_LSTM_selfAttention_model ]

check_point_file = [ ./Classification_LSTM_selfAttention/2018-12-04__05:33:09/model/ ]

-------------------------------------------------------------------------------------------

 ************************ Classification_LSTM_selfAttention_model structure ************************
< sequence_inputs > : 
Tensor("sequence_inputs:0", shape=(?, 200), dtype=int32)

< category_inputs > : 
Tensor("category_inputs:0", shape=(?,), dtype=int32)

< one_hot_category_inputs > : 
Tensor("one_hot:0", shape=(?, 6), dtype=float32)

< embeddings > : 
<tf.Variable 'embeddings:0' shape=(100000, 512) dtype=float32_ref>

< sequence_emb > : 
Tensor("sequence_emb/Identity:0", shape=(?, 200, 512), dtype=float32, device=/device:CPU:0)

< lstm_fw_cell > : 
<tensorflow.python.ops.rnn_cell_impl.LSTMCell object at 0x7f7e4a5e19e8>

< output_concat > : 
Tensor("Bi-LSTM/concat:0", shape=(?, 200, 64), dtype=float32)

< inner_W > : 
Tensor("self-attention/inner_tensordot:0", shape=(?, 200, 128), dtype=float32)

< outer_W > : 
<tf.Variable 'self-attention/outer_W:0' shape=(128, 64) dtype=float32_ref>

< inner_output > : 
Tensor("self-attention/inner_tensordot:0", shape=(?, 200, 128), dtype=float32)

< outer_output > : 
Tensor("self-attention/outer_tensordot:0", shape=(?, 200, 64), dtype=float32)

< alpha > : 
Tensor("self-attention/alpha:0", shape=(?, 200, 64), dtype=float32)

< attention_output > : 
Tensor("self-attention/output:0", shape=(?, 64, 64), dtype=float32)

< attention_flatten > : 
Tensor("self-attention/output_flatten/Reshape:0", shape=(?, 4096), dtype=float32)

< AA_T > : 
Tensor("attention-penalty/AA_T:0", shape=(?, 64, 64), dtype=float32)

< I > : 
Tensor("attention-penalty/I/MatrixDiag:0", shape=(64, 64), dtype=float32)

< penalty > : 
Tensor("attention-penalty/penalty:0", shape=(?,), dtype=float32)

< after_W > : 
<tf.Variable 'after-attention-fc/W:0' shape=(4096, 128) dtype=float32_ref>

< fc_ouptut > : 
Tensor("after-attention-fc/output:0", shape=(?, 128), dtype=float32)

< W_output > : 
<tf.Variable 'output/W_output:0' shape=(128, 6) dtype=float32_ref>

< b_output > : 
<tf.Variable 'output/b_output:0' shape=(6,) dtype=float32_ref>

< scores > : 
Tensor("output/scores:0", shape=(?, 6), dtype=float32)

< final_loss > : 
Tensor("loss/add:0", shape=(), dtype=float32)

< predictions > : 
Tensor("output/predictions:0", shape=(?,), dtype=int64)

< correct_predictions > : 
Tensor("accuracy/Equal:0", shape=(?,), dtype=bool)

< accuracy > : 
Tensor("accuracy/accuracy:0", shape=(), dtype=float32)

< train_final_loss_summary > : 
Tensor("summary/train_final_loss:0", shape=(), dtype=string)

< train_acc_summary > : 
Tensor("summary/train_accuracy:0", shape=(), dtype=string)

< lr_summary > : 
Tensor("summary/learning_rate:0", shape=(), dtype=string)

< train_summary_op > : 
Tensor("summary/Merge/MergeSummary:0", shape=(), dtype=string)

< test_summary_op > : 
Tensor("summary/Merge_1/MergeSummary:0", shape=(), dtype=string)

< gradients > : 
[<tensorflow.python.framework.ops.IndexedSlices object at 0x7f7e47481ac8>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_1:0' shape=(544, 128) dtype=float32>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_2:0' shape=(128,) dtype=float32>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_3:0' shape=(544, 128) dtype=float32>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_4:0' shape=(128,) dtype=float32>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_5:0' shape=(64, 128) dtype=float32>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_6:0' shape=(128, 64) dtype=float32>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_7:0' shape=(4096, 128) dtype=float32>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_8:0' shape=(128,) dtype=float32>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_9:0' shape=(128, 6) dtype=float32>, <tf.Tensor 'optimizer/clip_by_global_norm/optimizer/clip_by_global_norm/_10:0' shape=(6,) dtype=float32>]

< optimizer > : 
<tensorflow.python.training.adam.AdamOptimizer object at 0x7f7e47e22390>

< train_op > : 
name: "optimizer/Adam"
op: "NoOp"
input: "^optimizer/Adam/update_embeddings/group_deps"
input: "^optimizer/Adam/update_BiLSTM/fw/LSTM-forward/kernel/ApplyAdam"
input: "^optimizer/Adam/update_BiLSTM/fw/LSTM-forward/bias/ApplyAdam"
input: "^optimizer/Adam/update_BiLSTM/bw/LSTM-backward/kernel/ApplyAdam"
input: "^optimizer/Adam/update_BiLSTM/bw/LSTM-backward/bias/ApplyAdam"
input: "^optimizer/Adam/update_self-attention/inner_W/ApplyAdam"
input: "^optimizer/Adam/update_self-attention/outer_W/ApplyAdam"
input: "^optimizer/Adam/update_after-attention-fc/W/ApplyAdam"
input: "^optimizer/Adam/update_after-attention-fc/b/ApplyAdam"
input: "^optimizer/Adam/update_output/W_output/ApplyAdam"
input: "^optimizer/Adam/update_output/b_output/ApplyAdam"
input: "^optimizer/Adam/Assign"
input: "^optimizer/Adam/Assign_1"

# train accuracy, loss summary (1 epoch 당 약 1000개의 batch)

<img width="1616" alt="train" src="https://user-images.githubusercontent.com/37528988/49430663-f89b3b00-f7ee-11e8-95a0-7e95eaad617b.png">

# test accuracy, loss summary (1 epoch 당 약 280개의 batch)

<img width="1467" alt="test" src="https://user-images.githubusercontent.com/37528988/49430679-018c0c80-f7ef-11e8-92fe-8980a53cc1d4.png">

# 그래프

## Embedding

<img width="780" alt="embedding" src="https://user-images.githubusercontent.com/37528988/49430393-3fd4fc00-f7ee-11e8-8c47-5c35bbb93546.png">

## after Embedding

<img width="821" alt="default" src="https://user-images.githubusercontent.com/37528988/49430406-46637380-f7ee-11e8-9af0-2d5146c53f65.png">

## Bi-LSTM

<img width="1311" alt="bi-lstm2" src="https://user-images.githubusercontent.com/37528988/49430416-4e231800-f7ee-11e8-9a10-c7b7bb5de578.png">

## self-attention

<img width="985" alt="self attention" src="https://user-images.githubusercontent.com/37528988/49430419-524f3580-f7ee-11e8-9c73-f626d0999c68.png">

## optimizer

<img width="1014" alt="optimizer2" src="https://user-images.githubusercontent.com/37528988/49430436-5713e980-f7ee-11e8-97b5-8e408513b8c6.png">
